In [94]:
toronto_dropna = toronto_merged.dropna(subset=['Cluster Labels'])
toronto_d.shape

(103, 16)

# Segmenting and Clustering Neighborhoods in Toronto
## Week 3 Peer-graded Assignment

## Part 1
### Load the data from wikipedia webpage

In [67]:
import pandas as pd
import numpy as np

In [68]:
# in case lxml is not installed
# !pip install lxml

In [69]:
# load first table into dataframe
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [70]:
df.shape

(180, 3)

### Preprocess

In [71]:
df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

In [72]:
df['Neighbourhood'].value_counts()

Not assigned                                                                                                                              77
Downsview                                                                                                                                  4
Don Mills                                                                                                                                  2
Rosedale                                                                                                                                   1
Old Mill South, King's Mill Park, Sunnylea, Humber Bay, Mimico NE, The Queensway East, Royal York South East, Kingsway Park South East     1
                                                                                                                                          ..
Cedarbrae                                                                                                                                  1
Central Bay S

There is value with 'Not assigned' 77 each in both 'Borough' and 'Neighbourhood' columns

In [73]:
# replace 'Not assigned' as NaN
df = df.replace(to_replace='Not assigned', value=np.NaN)
df.head()

Postal Code           Borough              Neighbourhood
0         M1A               NaN                        NaN
1         M2A               NaN                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [74]:
df.isna().value_counts()

Postal Code  Borough  Neighbourhood
False        False    False            103
             True     True              77
dtype: int64

In [75]:
# drop record with NaN in 'Borough'
df = df.dropna(subset=['Borough'])
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [76]:
df.isna().value_counts()

Postal Code  Borough  Neighbourhood
False        False    False            103
dtype: int64

now, there is no NaN in both 'Borough' and 'Neighbourhood'

In [77]:
# check for duplicated postal codes
df['Postal Code'].value_counts().sort_values(ascending=False)

M9W    1
M4X    1
M4L    1
M6S    1
M3B    1
      ..
M5S    1
M2R    1
M3A    1
M5V    1
M4H    1
Name: Postal Code, Length: 103, dtype: int64

No duplicated postal code found.

In [78]:
df = df.reset_index(drop=True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [79]:
df.shape

(103, 3)

got the dataframe of 103 rows with 'Postal Code', 'Borough', 'Neighbourhood'

## Part 2

### Load geo data via geocoder
i tried to run the command, and it take forever.  
so skip this part to merge from .csv

In [80]:
# install geocoder
# !pip install geocoder

In [81]:
# import geocoder # import geocoder

# # initialize your variable to None
# lat_lng_coords = None
# postal_code = 'M3A'
# # loop until you get the coordinates
# while(lat_lng_coords is None):
#   g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#   lat_lng_coords = g.latlng

# latitude = lat_lng_coords[0]
# longitude = lat_lng_coords[1]

### Merge lat, lng data from Geospatial_Coordinates.csv

In [82]:
df_geo = pd.read_csv('Geospatial_Coordinates.csv')
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [83]:
df_full = pd.merge(df, df_geo, how='left')
df_full.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [84]:
df_full.shape

(103, 5)

got the dataframe of 103 rows with additional Latitude and Longitude data

## Part 3

In [85]:
import folium
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import requests
import os
from dotenv import load_dotenv

load_dotenv()

True

In [86]:
# df_toronto = df_full[df_full['Borough'].str.contains('toronto',case=False)]
# df_toronto.reset_index(drop=True).head()

In [87]:
df_map = df_full

In [88]:
# create map of Toronto using latitude and longitude values
latitude = 43.7100
longitude = -79.3832
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_map['Latitude'], df_map['Longitude'], df_map['Borough'], df_map['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Define Foursquare Credentials and Version

In [89]:
CLIENT_ID = os.getenv('CLIENT_ID') # your Foursquare ID
CLIENT_SECRET = os.getenv('CLIENT_SECRET') # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

### Create function to get nearby venues in Toronto

In [90]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [91]:
toronto_venues = getNearbyVenues(names=df_map['Neighbourhood'],
                                latitudes=df_map['Latitude'],
                                longitudes=df_map['Longitude']
                                )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

KeyError: 'groups'

In [26]:
# check size of resulting dataframe
print(toronto_venues.shape)
toronto_venues.head()

(2120, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

In [27]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             4   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    21   
...                                                                 ...   
Willowdale, Willowdale West                                           6   
Woburn                                                                4   
Woodbine Heights                                                      6   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              4   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     21   
...                                                                  ...   
Willowdale, Willowdale West                                            6   
Woburn                                                                 4   
Woodbine Heights                                                       6   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            4               4   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   21              21   
...                                                ...             ...   
Willowdale, Willowdale West                          6               6   
Woburn                                               4               4   
Woodbine Heights                                     6               6   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      4   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             21   
...                                                          ...   
Willowdale, Willowdale West                                    6   
Woburn                                                         4   
Woodbine Heights                                               6   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     4  
Alderwood, Long Branch                

In [28]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


### Analyze Each Neighbourhood

In [58]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  \
0                    0             0  ...           0   
1                    0             0  ...           0   
2                    0             0  ...           0   
3                    0             0  ...           0   
4                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 268 columns]

In [59]:
toronto_onehot.shape

(2120, 268)

### group rows by neighbourhood and taking mean of the frequency of each category

In [60]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
91                      Willowdale, Willowdale West                0.0   
92                                           Woburn                0.0   
93                                 Woodbine Heights                0.0   
94                                  York Mills West                0.0   
95                         York Mills, Silver Hills                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
1       0.0                 0.0           0.0             0.0   
2       0.0                 0.0           0.0             0.0   
3       0.0                 0.0           0.0             0.0   
4       0.0                 0.0           0.0             0.0   
..      ...                 ...           ...             ...   
91      0.0                 0.0           0.0             0.0   
92      0.0                 0.0           0.0             0.0   
93      0.0                 0.0           0.0             0.0   
94      0.0                 0.0           0.0             0.0   
95      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.047619           0.0  ...   
..              ...               ...                  ...           ...  ...   
91              0.0               0.0             0.000000           0.0  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   
95              0.0               0.0             0.000000           0.0  ...   

    Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0          0.0                            0.0               0.0   
1          0.0                            0.0               0.0   
2          0.0                            0.0               0.0   
3          0.0                            0.0               0.0   
4          0.0                            0.0               0.0   
..         ...                            ...               ...   
91         0.0                            0.0               0.0   
92         0.0                            0.0               0.0   
93         0.0                            0.0               0.0   
94         0.0                            0.0               0.0   
95         0.0                            0.0               0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0       0.0        0.0          0.0   
1                     0.0              0.0       0.0        0.0          0.0   
2                     0.0              0.0       0.0        0.0          0.0   
3                     0.0              0.0       0.0        0.0          0.0   
4                     0.0              0.0       0.0        0.0          0.0   
..                    ...              ...       ...        ...          ...   
91                    0.0     

In [61]:
toronto_grouped.shape

(96, 268)

#### Put data into pandas dataframe

In [62]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [63]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant    Chinese Restaurant   
1                   Gym             Sandwich Place           Coffee Shop   
2                  Bank                Supermarket      Sushi Restaurant   
3   Japanese Restaurant                       Bank    Chinese Restaurant   
4    Italian Restaurant                Coffee Shop      Greek Restaurant   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio                Drugstore   Distribution Center   
1          Skating Rink                      Pub               Dog Run   
2          Intersection            Deli / Bodega            Restaurant   
3           Yoga Studio      Distribution Center               Dog Run   
4       Thai Restaurant  Comfort Food Restaurant             Juice Bar   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0               Dog Run           Doner Restaurant             Donut Shop  
1    Dim Sum Restaurant                      Diner         Discount Store  
2        Ice Cream Shop  Middle Eastern Restaurant    Fried Chicken Joint  
3      Doner Restaurant                 Donut Shop              Drugstore  
4               Butcher                       Café             Restaurant

### Cluster Neighbourhoods
Run k-means to cluster the neighborhood into 5 clusters

In [64]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

Create new dataframe includes the cluster with top 10 venues for each neighborhood

In [65]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_map

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighbourhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             0.0      Food & Drink Shop   
1  43.725882 -79.315572             1.0  Portuguese Restaurant   
2  43.654260 -79.360636             1.0            Coffee Shop   
3  43.718518 -79.464763             1.0         Clothing Store   
4  43.662301 -79.389494             1.0            Coffee Shop   

  2nd Most Common Venue        3rd Most Common Venue   4th Most Common Venue  \
0                  Park  Eastern European Restaurant          Discount Store   
1           Coffee Shop            French Restaurant            Hockey Arena   
2                  Park                       Bakery                    Café   
3     Accessories Store                     Boutique  Furniture / Home Store   
4      Sushi Restaurant                  Yoga Studio                Creperie   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Distribution Center               Dog Run      Doner Restaurant   
1          Intersection            Donut Shop                 Diner   
2        Breakfast Spot                   Pub               Theater   
3           Event Space           Coffee Shop         Women's Store   
4              Beer Bar         Smoothie Shop        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Donut Shop             Drugstore            Yoga Studio  
1         Discount Store   Distribution Center                Dog Run  
2            Yoga Studio               Brewery             Shoe Store  
3  Vietnamese Restaurant   Filipino Restaurant                  Field  
4          Burrito Place                  Café  Portuguese Restaurant

In [102]:
toronto_merged[toronto_merged['Cluster Labels'].isna()]

Postal Code      Borough  \
5          M9A    Etobicoke   
11         M9B    Etobicoke   
95         M1X  Scarborough   

                                        Neighbourhood   Latitude  Longitude  \
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242   
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724   
95                                        Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5              NaN                   NaN                   NaN   
11             NaN                   NaN                   NaN   
95             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
5                    NaN                    NaN  
11                   NaN                    NaN  
95                   NaN                    NaN

In [101]:
toronto_clean = toronto_merged.dropna(subset=['Cluster Labels'])
toronto_clean['Cluster Labels'].isna().value_counts()

False    100
Name: Cluster Labels, dtype: int64

In [104]:
toronto_clean = toronto_clean.astype({'Cluster Labels':'int'})
toronto_clean.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656               0      Food & Drink Shop   
1  43.725882 -79.315572               1  Portuguese Restaurant   
2  43.654260 -79.360636               1            Coffee Shop   
3  43.718518 -79.464763               1         Clothing Store   
4  43.662301 -79.389494               1            Coffee Shop   

  2nd Most Common Venue        3rd Most Common Venue   4th Most Common Venue  \
0                  Park  Eastern European Restaurant          Discount Store   
1           Coffee Shop            French Restaurant            Hockey Arena   
2                  Park                       Bakery                    Café   
3     Accessories Store                     Boutique  Furniture / Home Store   
4      Sushi Restaurant                  Yoga Studio                Creperie   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Distribution Center               Dog Run      Doner Restaurant   
1          Intersection            Donut Shop                 Diner   
2        Breakfast Spot                   Pub               Theater   
3           Event Space           Coffee Shop         Women's Store   
4              Beer Bar         Smoothie Shop        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Donut Shop             Drugstore            Yoga Studio  
1         Discount Store   Distribution Center                Dog Run  
2            Yoga Studio               Brewery             Shoe Store  
3  Vietnamese Restaurant   Filipino Restaurant                  Field  
4          Burrito Place                  Café  Portuguese Restaurant

#### Finally, visualize the resulting clusters

In [105]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_clean['Latitude'], toronto_clean['Longitude'], toronto_clean['Neighbourhood'], toronto_clean['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

#### Cluster 0 : The Park

In [107]:
toronto_clean.loc[toronto_clean['Cluster Labels'] == 0, toronto_clean.columns[[1] + list(range(5, toronto_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York               0     Food & Drink Shop   
16              York               0                 Field   
21              York               0                  Park   
35         East York               0                  Park   
49        North York               0                  Park   
52        North York               0                  Park   
66        North York               0                  Park   
83   Central Toronto               0                  Park   
85       Scarborough               0          Intersection   
91  Downtown Toronto               0                  Park   
98         Etobicoke               0                  Park   

   2nd Most Common Venue        3rd Most Common Venue  \
0                   Park  Eastern European Restaurant   
16                 Trail                         Park   
21         Women's Store                         Pool   
35           Coffee Shop            Convenience Store   
49                Bakery   Construction & Landscaping   
52           Yoga Studio                    Drugstore   
66     Convenience Store                  Yoga Studio   
83                 Trail                 Tennis Court   
85                  Park                   Playground   
91            Playground                        Trail   
98                 River                         Pool   

      4th Most Common Venue        5th Most Common Venue  \
0            Discount Store          Distribution Center   
16             Hockey Arena                        Diner   
21              Yoga Studio                   Donut Shop   
35              Yoga Studio                    Drugstore   
49              Yoga Studio  Eastern European Restaurant   
52                    Diner               Discount Store   
66                Drugstore               Discount Store   
83  Comfort Food Restaurant                   Comic Shop   
85               Donut Shop           Dim Sum Restaurant   
91         Doner Restaurant           Dim Sum Restaurant   
98         Doner Restaurant           Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Dog Run      Doner Restaurant            Donut Shop   
16        Discount Store   Distribution Center               Dog Run   
21                 Diner        Discount Store   Distribution Center   
35        Discount Store   Distribution Center               Dog Run   
49   Distribution Center               Dog Run      Doner Restaurant   
52   Distribution Center               Dog Run      Doner Restaurant   
66   Distribution Center               Dog Run      Doner Restaurant   
83    Falafel Restaurant           Event Space  Ethiopian Restaurant   
85                 Diner        Discount Store   Distribution Center   
91                 Diner        Discount Store   Distribution Center   
98                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue       10th Most Common Venue  
0              Drugstore                  Yoga Studio  
16      Doner Restaurant                   Donut Shop  
21               Dog Run             Doner Restaurant  
35      Doner Restaurant                   Donut Shop  
49            Donut Shop                    Drugstore  
52            Donut Shop  Eastern European Restaurant  
66            Donut Shop            Electronics Store  
83           Escape Room            Electronics Store  
85               Dog Run             Doner Restaurant  
91               Dog Run                   Donut Shop  
98               Dog Run                   Donut Shop

#### Cluster 1: Coffee shope, Café?

In [109]:
toronto_clean.loc[toronto_clean['Cluster Labels'] == 1, toronto_clean.columns[[1] + list(range(5, toronto_clean.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
1          North York               1  Portuguese Restaurant   
2    Downtown Toronto               1            Coffee Shop   
3          North York               1         Clothing Store   
4    Downtown Toronto               1            Coffee Shop   
7          North York               1                    Gym   
..                ...             ...                    ...   
97   Downtown Toronto               1            Coffee Shop   
99   Downtown Toronto               1            Coffee Shop   
100      East Toronto               1     Light Rail Station   
101         Etobicoke               1          Deli / Bodega   
102         Etobicoke               1         Hardware Store   

    2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
1             Coffee Shop     French Restaurant            Hockey Arena   
2                    Park                Bakery                    Café   
3       Accessories Store              Boutique  Furniture / Home Store   
4        Sushi Restaurant           Yoga Studio                Creperie   
7              Restaurant            Beer Store     Japanese Restaurant   
..                    ...                   ...                     ...   
97                   Café                 Hotel     Japanese Restaurant   
99       Sushi Restaurant   Japanese Restaurant              Restaurant   
100           Yoga Studio         Garden Center              Skate Park   
101        Baseball Field           Yoga Studio               Drugstore   
102         Tanning Salon         Grocery Store    Fast Food Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1            Intersection            Donut Shop                 Diner   
2          Breakfast Spot                   Pub               Theater   
3             Event Space           Coffee Shop         Women's Store   
4                Beer Bar         Smoothie Shop        Sandwich Place   
7             Coffee Shop  Caribbean Restaurant         Shopping Mall   
..                    ...                   ...                   ...   
97                    Gym            Restaurant           Salad Place   
99                Gay Bar  Fast Food Restaurant           Men's Store   
100        Farmers Market  Fast Food Restaurant         Burrito Place   
101   Distribution Center               Dog Run      Doner Restaurant   
102        Discount Store     Convenience Store         Burrito Place   

        8th Most Common Venue        9th Most Common Venue  \
1              Discount Store          Distribution Center   
2                 Yoga Studio                      Brewery   
3       Vietnamese Restaurant          Filipino Restaurant   
4               Burrito Place                         Café   
7                        Café             Sushi Restaurant   
..                        ...                          ...   
97           Asian Restaurant                   Steakhouse   
99   Mediterranean Restaurant                         Café   
100                Restaurant                Auto Workshop   
101                Donut Shop  Eastern European Restaurant   
102            Sandwich Place                 Burger Joint   

    10th Most Common Venue  
1                  Dog Run  
2               Shoe Store  
3                    Field  
4    Portuguese Restaurant  
7              Supermarket  
..                     ...  
97     American Restaurant  
99             Yoga Studio  
100                 Garden  
101                  Diner  
102        Supplement Shop  

[86 rows x 12 columns]

### Outliers

#### Cluster 2

In [110]:
toronto_clean.loc[toronto_clean['Cluster Labels'] == 2, toronto_clean.columns[[1] + list(range(5, toronto_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York               2   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45    Dim Sum Restaurant    Falafel Restaurant           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45  Ethiopian Restaurant           Escape Room     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
45  Eastern European Restaurant              Drugstore

#### Cluster 3

In [111]:
toronto_clean.loc[toronto_clean['Cluster Labels'] == 3, toronto_clean.columns[[1] + list(range(5, toronto_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
64    York               3         Jewelry Store           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
64             Drugstore        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
64               Dog Run      Doner Restaurant            Donut Shop   

          9th Most Common Venue 10th Most Common Venue  
64  Eastern European Restaurant                  Field

#### Cluster 4

In [112]:
toronto_clean.loc[toronto_clean['Cluster Labels'] == 4, toronto_clean.columns[[1] + list(range(5, toronto_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough               4  Fast Food Restaurant    Dim Sum Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6        Farmers Market    Falafel Restaurant           Event Space   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Ethiopian Restaurant           Escape Room     Electronics Store   

         9th Most Common Venue 10th Most Common Venue  
6  Eastern European Restaurant              Drugstore